src/Symfony/Component/Messenger/Middleware/NextMiddleware.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Messenger\Envelope;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class NextMiddleware implements MiddlewareInterface, NextInterface
+{
+    private $middlewareIterator;
+
+    public function __construct(\Iterator $middlewareIterator = null)
+    {
+        $this->middlewareIterator = $middlewareIterator;
+    }
+
+    public function __invoke(Envelope $envelope = null): MiddlewareInterface
+    {
+        if (null === $iterator = $this->middlewareIterator) {
+            return $this;
+        }
+        $iterator->next();
+
+        if (!$iterator->valid()) {
+            $this->middlewareIterator = null;
+
+            return $this;
+        }
+        $middleware = $iterator->current();
+
+        if (null !== $envelope) {
+            $middleware->handle($envelope, $this);
+        }
+
+        return $middleware;
+    }
+
+    public function handle(Envelope $envelope, callable $next): void
```

Comment:
why should this class implement MiddlewareInterface ? Calling it as a middleware does not make sense